src/prepared_statement.c

```diff
@@ -0,0 +1,388 @@
+	{
+		/* Maintain LRU double-linked list: move element to the end of the list */
+		HASH_UNLINK(hh, server->server_prepared_statements, server_ps);
+		HASH_APPEND_LIST(hh, server->server_prepared_statements, server_ps);
+	}
+
+	slog_debug(client, "handle_bind_command: mapped statement '%s' (query '%s') to '%s'",
+			   pp.name, ps->query_with_params, ps->stmt_name);
+
+	diff = strlen(bp.name) - ps->stmt_name_len;
+	len = pkt->len - diff;
+	buf = pktbuf_dynamic(len);
+	pktbuf_put_char(buf, pkt->type);
+	pktbuf_put_uint32(buf, len - 1); /* length does not include type byte */
+	pktbuf_put_string(buf, bp.portal);
+	pktbuf_put_string(buf, ps->stmt_name);
+	pktbuf_put_bytes(buf, pkt->data.data + buf->write_pos + diff, len - buf->write_pos);
+	if (!pktbuf_send_queued(buf, server)) {
+		pktbuf_free(buf);
+		return false;
+	}
+	return true;
+}
+
+bool handle_describe_command(PgSocket *client, PktHdr *pkt)
+{
+	SBuf *sbuf = &client->sbuf;
+	PgSocket *server = client->link;
+	PgDescribePacket dp;
+	PgClientPreparedStatement *client_ps = NULL;
+	PgServerPreparedStatement *server_ps = NULL;
+	PgPreparedStatement* ps;
+	PktBuf *buf;
+	bool res;
+
+	Assert(server);
+
+	if (!unmarshall_describe_packet(client, pkt, &dp) || dp.type != 'S')
+		return false;
+
+	HASH_FIND_STR(client->client_prepared_statements, dp.name, client_ps);
+	if (!client_ps) {
+		disconnect_client(client, true, "prepared statement '%s' not found", dp.name);
+		return false;
+	}
+	ps = client_ps->ps;
+
+	HASH_FIND_INT(server->server_prepared_statements, &ps->query_id, server_ps);
+	if (!server_ps) {
+		disconnect_client(client, true, "prepared statement '%s' not found at server", dp.name);
+		return false;
+	}
+	sbuf_prepare_skip(sbuf, pkt->len);
+	if (!sbuf_flush(sbuf))
+		return false;
+
+	slog_debug(client, "handle_describe_command: mapped statement '%s' (query '%s') to '%s'",
+			   dp.name, ps->query_with_params, ps->stmt_name);
+
+	buf = create_describe_packet(ps->stmt_name);
+
+	res = pktbuf_send_immediate(buf, server);
+	pktbuf_free(buf);
+	return res;
+}
+
+bool handle_close_statement_command(PgSocket *client, PktHdr *pkt, PgClosePacket *close_packet)
+{
+	SBuf *sbuf = &client->sbuf;
+	PgClientPreparedStatement *client_ps = NULL;
+	PktBuf *buf;
+	bool res = true;
+
+	HASH_FIND_STR(client->client_prepared_statements, close_packet->name, client_ps);
+	if (client_ps) {
+		slog_noise(client, "handle_close_command: removed '%s' from cached prepared statements, items remaining %u", close_packet->name, HASH_COUNT(client->client_prepared_statements));
+		HASH_DEL(client->client_prepared_statements, client_ps);
+		if (--client_ps->ps->use_count == 0)
+		{
+			HASH_DEL(prepared_statements, client_ps->ps);
+			free(client_ps->ps);
+		}
+		free(client_ps);
+
+		/* Do not forward packet to server */
+		sbuf_prepare_skip(sbuf, pkt->len);
+		if (!sbuf_flush(sbuf))
+			return false;
+
+		buf = create_close_complete_packet();
+		res = pktbuf_send_immediate(buf, client);
+		pktbuf_free(buf);
+	}
```

Comment:
Why?
If prepared statement is unknown we can just send this command to the server and let it react o this command.

Reply:
There's no real point in doing so, because the statement names clients use match are different from the ones that we preprae on the server. So best case: it causes extra latency. Worst case: their statement name actually conflicts with some of our own prepared statement names and an unintendedly closes a completely unrelated prepared statement (quite unlikely but theoretically possible).

Reply:
Sorry, I do not agree.
First of all, if it is possible not to do something and redirect work to somebody else, then it should be done:)
More seriously: it is up to postgres how to react on close command for the statement which was not prepared or already closed. We should not try to simulate this behavior, because it can be changed. And there is IMHO no sense for it - it is in any case "error path". So  latency is not critical here.